### ISSUE TITLE
Broken Source Maps or Stack Traces for async functions after upgrading from 3.1.0 to 3.1.1

### ISSUE THREAD

**Issue by donaldpipowitch** [mochajs/mocha#2528]
Given a test like this:

```typescript
import 'babel-polyfill';
import expect from 'expect';

describe('test my code', () => {
  it('should show my app', async (): Promise<any> => {
    expect(2).toBe(1);
  });
});
```

`3.1.1` shows:

```
ERROR: 'Unhandled promise rejection', TypeError{stack: 'fail@http://localhost:9876/absolute/Users/foo/Workspace/project/node_modules/mocha/mocha.js?60c338ed8db7987f07880c121c83fed2648782f6:4726:6
```

And also:

```
10 10 2016 15:35:06.809:WARN [PhantomJS 2.1.1 (Mac OS X 0.0.0)]: Disconnected (1 times), because no message in 10000 ms.
PhantomJS 2.1.1 (Mac OS X 0.0.0) ERROR
      Disconnected, because no message in 10000 ms.
```

`3.1.0` correctly reports an error with a correct stack trace and isn't disconnected from PhantomJS. (I guess the promise rejection is handled correctly.)

**Note** that I use an `async` function here. (In my real test I also `await` something, but it is not necessary to do that to reproduce the error.

Maybe it was introduced by fixing https://github.com/mochajs/mocha/issues/1417?

**Comment by boneskull** [comment]
I'm going to guess this is Karma, karma-mocha, karma-babel-preprocessor and karma-phantomjs-launcher... but really could use some more info

**Comment by boneskull** [comment]
wait typescript?

**Comment by donaldpipowitch** [comment]
Yes to everything, except karma-babel-preprocessor ;)

**Comment by boneskull** [comment]
I can trivially reproduce with just karma, karma-mocha and karma-phantomjs-launcher (phantomjs 2.1.1)

**Comment by boneskull** [comment]
okay

**Comment by boneskull** [comment]
but not 100% that this is a bug yet

**Comment by boneskull** [comment]
okay, yeah this looks like a failure to properly handle the rejection.  fwiw, the previous behavior was *also* incorrect, even though mocha was able to handle it.

**Comment by boneskull** [comment]
alright, I got something

**Comment by donaldpipowitch** [comment]
Cool! Thank you for the quick response.

**Comment by boneskull** [comment]
np.  we don't have the greatest browser coverage, so this is bound to happen from time to time.

published as [v3.1.2](https://git.io/vPE4D)
